src/VisualStudio/Core/Def/Implementation/Snippets/AbstractSnippetExpansionClient.cs

```diff
@@ -403,7 +772,12 @@ public int EndExpansion()
                 earlyEndExpansionHappened = true;
             }
 
-            ExpansionSession = null;
+            // This call to EndExpansion may occur within a call to InsertSpecificExpansion (the current snippet session
+            // is terminated before the new one is created to replace it). Since _state may contain symbol information
+            // used by snippet functions during the creation of the new snippet, we only want to clear symbol
+            // information if the state hasn't set the _preserveSymbols flag.
```

Comment:
seems really complex.  could you clarify why you can't o this route:

 you know teh new expansion is being created.  At that point that code grabs the current state and hten manipulates it directly to get hte 'new' state and then just sets that...

then, it doesn't matter what this code does as the outer code just slams in whatever state it wants over the current state...

(i could totally be missing something).

Reply:
The snippet client is reentrant. We call the editor, and then the editor calls the following in sequence _before_ returning to this method:

1. Clear the current session
2. Get the values for the new snippet

Without the manual state tracking, (1) would invalidate the data we need for a reentrant operation (2).

Reply:
⏱ updated documentation